Add Barotropic2D EOS

nilsdeppe opened the pull request:
## Proposed changes

Will be used in Newtonian Euler so that the EOS can be runtime and is always 2d.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add Barotropic2D equation of state

nilsdeppe committed:
Add promote_to_2d_eos function to 1d EOSes

kidder commented on src/PointwiseFunctions/Hydro/EquationsOfState/Barotropic2D.hpp:
```diff
@@ -0,0 +1,176 @@
+ * The equation of state takes the form
+ *
+ * \f[
+ * p = p (T, rho, Y_e) = p(0, rho, Y_e= Y_{e, \beta})
```
fix the equation (rho -> \rho)

kidder commented on src/PointwiseFunctions/Hydro/EquationsOfState/Barotropic2D.hpp:
```diff
@@ -0,0 +1,176 @@
+ * \f]
+ *
+ * where \f$\rho\f$ is the rest mass density, \f$T\f$  the
+ * temperatur , and \f$Y_e\f$ is the electron fraction are not
```
temperature

kidder replied:
electron fraction.  The temperature and electron fraction are not used, so evaluating this EoS at any arbitrary temperature or electron fraction is...

kidder commented on src/PointwiseFunctions/Hydro/EquationsOfState/Barotropic2D.hpp:
```diff
@@ -0,0 +1,176 @@
+ * temperatur , and \f$Y_e\f$ is the electron fraction are not
+ * used, and therefore this evaluating this EoS at any arbtirary
+ * temeperature or electron fraction is equivalent to evaluating it at
+ * temperature 0 and in beta equalibrium
```
zero temperature

kidder commented on src/PointwiseFunctions/Hydro/EquationsOfState/Barotropic2D.hpp:
```diff
@@ -0,0 +1,176 @@
+    using type = ColdEos;
+    static std::string name() { return pretty_type::short_name<ColdEos>(); }
+    static constexpr Options::String help{
+        "The underlying Eos which is being represented as a "
```
Eos

kidder commented on src/PointwiseFunctions/Hydro/EquationsOfState/Barotropic2D.hpp:
```diff
@@ -0,0 +1,176 @@
+    static std::string name() { return pretty_type::short_name<ColdEos>(); }
+    static constexpr Options::String help{
+        "The underlying Eos which is being represented as a "
+        "2D Eos.  Must be a 1D EoS"};
```
EoS

kidder commented on src/PointwiseFunctions/Hydro/EquationsOfState/Barotropic2D.hpp:
```diff
@@ -0,0 +1,176 @@
+  Barotropic2D& operator=(const Barotropic2D&) = default;
+  Barotropic2D(Barotropic2D&&) = default;
+  Barotropic2D& operator=(Barotropic2D&&) = default;
+  ~Barotropic2D() override = default;
```
clang-tidy wants this to be virtual

nilsdeppe replied:
I think this is a clang-tidy bug. `override` implies virtual

kidder commented on tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Test_PiecewisePolytropicFluid.cpp:
```diff
@@ -342,6 +343,9 @@ SPECTRE_TEST_CASE(
   // Check the correct 3D EoS is got
   CHECK(*eos.promote_to_3d_eos() ==
         EoS::Barotropic3D<EoS::PiecewisePolytropicFluid<true>>(eos));
+  // Check the correct 2D EoS is got
```
is obtained

kidder reviewed (changes requested):
you can squash the changes

nilsdeppe committed:
Fix typos in Barotropic3D EOS

nilsdeppe commented:
I addressed the comments (hopefully fully :D ) and just squashed directly. I also added a commit that makes the same fixes to the Barotropic3D eos. Thanks for the review!
